src/library/scala/concurrent/duration/Duration.scala

```diff
@@ -124,6 +124,8 @@ object Duration {
       Undefined
     else if (nanos > Long.MaxValue || nanos < Long.MinValue)
       throw new IllegalArgumentException("trying to construct too large duration with " + nanos + "ns")
+    else if (nanos < 0)
+      fromNanos((nanos - 0.5).toLong)
```

Comment:
If we want to truncate, we shouldn't do operations on a double: it's probably the reason for the loss of precision.
I personally don't think `0.5ns` should be rounded to `1ns`, but we could use `math.rint` instead, if we want to round, i.e.
```scala
else
  fromNanos(math.rint(nanos).toLong)
```
which works for negatives the same way as for positives.